[db][protocol] Implement TeamSubscription2 DB shapes and migration

jankeromnes opened the pull request:
## Description
<!-- Describe your changes in detail -->

Implement TeamSubscription2 DB shapes and migration.

Part of https://github.com/gitpod-io/gitpod/issues/7759, split out of https://github.com/gitpod-io/gitpod/pull/8041

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/9652

## How to test
<!-- Provide steps to test this PR -->

1. Should build
2. DB should contain new table `d_b_team_subscription2` + an additional column in both `d_b_subscription` and `d_b_team_membership` tables

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

jankeromnes commented:
/werft run

AlexTugarev commented:
Looking into this now, and reading up the plan.

jankeromnes commented:
Many thanks @AlexTugarev! 😊 Let me know if you have any questions 🙏

jankeromnes commented:
Friendly ping @AlexTugarev 😇 🧡

AlexTugarev reviewed (commented):
`DBTeamMembership.subscriptionId` and `DBSubscription.teamMembershipId` seems to be redundant.

Looking at the usage in https://github.com/gitpod-io/gitpod/pull/8041/commits/81b41f18b0f00ce58bc8af4ed9c5ca76b5020c54#diff-349769928506cc98ef99e13b1d146ae991fd2755d6076cda2ed87b1693eec071R36-R37

What's the benefit?

jankeromnes commented:
Thanks for the feedback @AlexTugarev!

> `DBTeamMembership.subscriptionId` and `DBSubscription.teamMembershipId` seems to be redundant.
> [...]
> What's the benefit?

This is modelled on `DBTeamSubscriptionSlot.subscriptionId` and `DBSubscription.teamSubscriptionSlotId` being also redundant (or rather, they materialize a relationship in both directions), which has been helpful on several occasions (e.g. for bugs, or auditing purposes, when you lose one side of the link but not the other).

For example, when you leave a team, the membership is automatically deleted (leaving no trace), but the subscription is merely cancelled at end of term. This leaves a trace (e.g. for users wondering why they have unlimited hours without a paid subscription, we could see that the still-active subscription came from a team membership that has since been deleted).

AlexTugarev commented:
@jankeromnes, I find the cut of the PRs less useful for a review. I understand #8041 is a 🐋, but it seems cutting out the DB shapes doesn't much, as one still need to understand the call sites. 

How else do you expect this to be reviewed?

jankeromnes commented on components/gitpod-db/src/typeorm/migration/1650526577994-TeamSubscrition2.ts:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    public async down(queryRunner: QueryRunner): Promise<void> {}
```
Good catch! But, it should be pretty safe to omit this tear down, no? (I.e. up and down migrations can safely happen as many times and in any order as we like, the end result is still what we want.)

AlexTugarev replied:
In general, no. In case we'd need to revert, a following changed application is more complicated.

jankeromnes replied:
I'd prefer to leave this as is for now, but if we do need to permanently revert, I volunteer myself to drop the extra tables/columns manually.

jankeromnes replied:
Hmm, well okay, I might as well implement the tear down while we're here. 👍

AlexTugarev commented on components/gitpod-db/src/typeorm/team-db-impl.ts:
```diff
@@ -192,6 +197,21 @@ export class TeamDBImpl implements TeamDB {
+        if (!membership) {
+            throw new Error("The user is not currently a member of this team");
+        }
+        membership.subscriptionId = subscriptionId;
```
from #9655 I can read that the subscription already knows the membership. why storing this link in both directions?

jankeromnes replied:
See https://github.com/gitpod-io/gitpod/pull/9655#issuecomment-1120769840

AlexTugarev replied:
(Let's use this as a thread in here.)

> For example, when you leave a team, the membership is automatically deleted (leaving no trace), but the subscription is merely cancelled at end of term. This leaves a trace (e.g. for users wondering why they have unlimited hours without a paid subscription, we could see that the still-active subscription came from a team membership that has since been deleted).

So, what is the `DBTeamMembership.subscriptionId` need for then? I couldn't spot any usage.

jankeromnes replied:
Sure! So, while this direction of the link is indeed not used in the current code, it was requested in the RFC that this link be materialized in both directions:

> - We add a 0..1 reference from `d_b_team_member[ship]`  to `d_b_subscription` and the opposite reference in analogy to the reference to `d_b_team_subscription_slot`. 

[Source](https://www.notion.so/gitpod/Aligning-Team-Plans-and-Teams-9073db94359549918ab5c85cc364de46#e207bc6c7bf24b578fae8f7b9bd9d5d8) (internal)

And, this made sense to me at the time, because the link between `d_b_team_subscription_slot` and `d_b_subscription` also went both ways.

As mentioned above, this slight duplication has helped/saved us a few times, for example in customer requests, or to investigate recover from bugs.

Are you proposing to drop the reference from `d_b_team_membership` to `d_b_subscription` now?

AlexTugarev commented on components/gitpod-protocol/src/accounting-protocol.ts:
```diff
@@ -158,6 +159,15 @@ export namespace AssignedTeamSubscription {
+}
+export namespace AssignedTeamSubscription2 {
+    export function is(data: any): data is AssignedTeamSubscription2 {
+        return !!data && data.hasOwnProperty("teamMembershipId");
```
instead of `!data`, please use `typeof data === "object"` instead.

jankeromnes commented:
Addressed all nits, DB still looks good!

<details><summary>(see tables)</summary>

```
mysql> describe d_b_team_subscription2;
+------------------+--------------+------+-----+----------------------+--------------------------------+
| Field            | Type         | Null | Key | Default              | Extra                          |
+------------------+--------------+------+-----+----------------------+--------------------------------+
| id               | char(36)     | NO   | PRI | NULL                 |                                |
| teamId           | char(36)     | NO   | MUL | NULL                 |                                |
| paymentReference | varchar(255) | NO   |     | NULL                 |                                |
| startDate        | varchar(255) | NO   |     | NULL                 |                                |
| endDate          | varchar(255) | NO   |     |                      |                                |
| planId           | varchar(255) | NO   |     | NULL                 |                                |
| quantity         | int(11)      | NO   |     | NULL                 |                                |
| cancellationDate | varchar(255) | NO   |     |                      |                                |
| deleted          | tinyint(4)   | NO   |     | 0                    |                                |
| _lastModified    | timestamp(6) | NO   | MUL | CURRENT_TIMESTAMP(6) | on update CURRENT_TIMESTAMP(6) |
+------------------+--------------+------+-----+----------------------+--------------------------------+
10 rows in set (0.01 sec)

mysql> describe d_b_subscription;
+------------------------+--------------+------+-----+----------------------+--------------------------------+
| Field                  | Type         | Null | Key | Default              | Extra                          |
+------------------------+--------------+------+-----+----------------------+--------------------------------+
| id                     | int(11)      | YES  |     | NULL                 |                                |
| userId                 | char(36)     | NO   | MUL | NULL                 |                                |
| startDate              | varchar(255) | NO   |     | NULL                 |                                |
| endDate                | varchar(255) | NO   |     |                      |                                |
| amount                 | double       | NO   |     | NULL                 |                                |
| uid                    | char(36)     | NO   | PRI | NULL                 |                                |
| _lastModified          | timestamp(6) | NO   | MUL | CURRENT_TIMESTAMP(6) | on update CURRENT_TIMESTAMP(6) |
| planId                 | varchar(255) | NO   | MUL | free                 |                                |
| paymentReference       | varchar(255) | NO   |     |                      |                                |
| deleted                | tinyint(4)   | NO   |     | 0                    |                                |
| cancellationDate       | varchar(255) | NO   |     |                      |                                |
| paymentData            | text         | YES  |     | NULL                 |                                |
| teamSubscriptionSlotId | char(255)    | NO   | MUL |                      |                                |
| firstMonthAmount       | double       | YES  |     | NULL                 |                                |
| teamMembershipId       | char(36)     | NO   |     |                      |                                |
+------------------------+--------------+------+-----+----------------------+--------------------------------+
15 rows in set (0.05 sec)

mysql> describe d_b_team_membership;
+----------------+--------------+------+-----+----------------------+--------------------------------+
| Field          | Type         | Null | Key | Default              | Extra                          |
+----------------+--------------+------+-----+----------------------+--------------------------------+
| id             | char(36)     | NO   | PRI | NULL                 |                                |
| teamId         | char(36)     | NO   | MUL | NULL                 |                                |
| userId         | char(36)     | NO   | MUL | NULL                 |                                |
| role           | varchar(255) | NO   |     | NULL                 |                                |
| creationTime   | varchar(255) | NO   |     | NULL                 |                                |
| deleted        | tinyint(4)   | NO   |     | 0                    |                                |
| _lastModified  | timestamp(6) | NO   | MUL | CURRENT_TIMESTAMP(6) | on update CURRENT_TIMESTAMP(6) |
| subscriptionId | char(36)     | NO   |     |                      |                                |
+----------------+--------------+------+-----+----------------------+--------------------------------+
8 rows in set (0.02 sec)
```

</details>

AlexTugarev reviewed (approved):
LGTM!
